pymeasure/instruments/instrument.py

```diff
@@ -377,7 +377,24 @@ def fget(self,
                     )
             else:
                 vals = get_process(vals)
-                return vals
+                if not map_values:
+                    return vals
+                elif isinstance(values, (list, tuple, range)):
+                    return [values[int(val)] for val in vals]
+                elif isinstance(values, dict):
+                    ret = []
+                    for i, val in enumerate(vals):
+                        for k, v in values.items():
+                            if v == val:
+                                ret.append(k)
+                        if len(ret) != i + 1:
+                            raise KeyError(f"Value {val} not found in mapped values")
+                    return ret
+                else:
+                    raise ValueError(
+                        'Values of type `{}` are not allowed '
+                        'for Instrument.control'.format(type(values))
+                    )
```

Comment:
What I don't like is that this duplicates a lot of code and logic. I have come up with the following better alternative, which passes your test but does not duplicate the code. 
```
        def fget(self,
                 get_command=get_command,
                 values=values,
                 map_values=map_values,
                 get_process=get_process,
                 command_process=command_process,
                 check_get_errors=check_get_errors,
                 ):
            if get_command is None:
                raise LookupError("Instrument property can not be read.")
            vals = self.values(command_process(get_command), **kwargs)
            if check_get_errors:
                self.check_errors()
            vals = [get_process(v) for v in vals]  # TODO: should we run get_process on all elements or the whole list?
            if map_values:
                ret = []
                for value in vals:
                    if isinstance(values, (list, tuple, range)):
                        ret.append(values[int(value)])
                    elif isinstance(values, dict):
                        for k, v in values.items():
                            if v == value:
                                ret.append(k)
                                break
                        else:  # for-loop ran completely
                            raise KeyError(f"Value {value} not found in mapped values")
                    else:
                        raise ValueError(
                            'Values of type `{}` are not allowed '
                            'for Instrument.control'.format(type(values))
                        )
            else:
                ret = vals
            if len(ret) == 1:
                return ret[0]
            else:
                return ret
```
(basically, I am delaying the distinction between 1 and multiple value elements to later and do mapping on all elements)

However, I am not yet 100% confident that there are not further implications that I did not consider. 
* Also, I have not yet thought through at which point get_process should apply. 
* I'm not sure about mapping now becoming asymmetric, as mapping lists of items now only applies on get. Does it even make sense on set?

It would be good to have the extended amount of tests from #665 (will merge soon) before merging this.

ping @bmoneke what do you think about the impact of this?

Reply:
I'd say make `get_process(vals)`, because it gives flexibility without loosing possibility:

- If you want to modify every entry of a list, get_process can take that into account. e.g. `lambda vals: [val**2 for val in vals]`.
- If you want to take the list as a whole, e.g. get the average or standard deviation or sum or min, max or a specific value (e.g. the first one) etc., you can do it as well.

Reply:
I just noticed problem with my suggestion:
If you had `get_process = lambda v: 2*v` before, you need to work with the list now.

I do not know by heart how many instruments implement get_process.

On the other hand, applying the get process to every single value should not brake any present code.

Reply:
Good point on the flexibility!

Backwards compatibility is a good point (get_process is used a lot)! 
Until now, get_process is acting on the string entry if vals was length 1, and on the list if length >1 (which may be less than ideal in itself). We could preserve that behaviour with some small code tweaks.

Reply:
I used the code suggestion from above. Its true that is easier like you propose it. It seems to work for my device. 
I did not understand from your discussion if you still want to handle get_process differently?

Reply:
> I did not understand from your discussion if you still want to handle get_process differently?

> Event though it does not offer that much flexibility (e.g. averaging), I think your implementation is good, due to backwards compatibility.

> On the other hand, applying the get process to every single value should not brake any present code.

I think it's the other way round, we have changed behaviour of `get_process` when multiple elements are present:

> Until now, get_process is acting on the string entry if vals was length 1, and on the list if length >1 (which may be less than ideal in itself). We could preserve that behaviour with some small code tweaks.

I'll think on this, come up with a test and will suggest a tweak of that part of the code.

Reply:
OTOH, the docstring of `get_process` is "A function that take a value and allows processing before value mapping, returning the processed value", suggesting it acts on single values. 

In addition, we have the `preprocess_reply` that acts on the whole message (before it runs through `values()`, though)

Also, having  `get_process` always act on single elements simplifies the situation when the returned list can have a variable number of elements (but I don't know how frequent that is).

Reply:
@msmttchr @bmoneke what do you think the behaviour of `get_process()` with multiple values should be? Act on the whole list (like now, would make code slightly more complex, but is more flexible), or individual elements (fits smoother into the control flow, more consistent with single-element list use).
I'm on the fence, and my dice are a bad basis for decisions here. :grin:

Reply:
Sorry for not coming back. It was hard for me to decide, what to do. I definitively like the idea of doing some fancy method on the returned values list (e.g. sum and standard deviation...), but that would require to have a more complex `get_process = lambda v: 5 * v[0]` instead of just multiplying with some number.

Whenever you need complex mathematics on the return values (I'd say rarely), you can use a python property. In the many cases of a single return value, we should have a more simple `get_process`.

Therefore I propose that `get_process` acts on every single value.

Reply:
Thank you! I opened #740 for nailing down the current behaviour, integrating the PR, and writing docs and changelog for this change. Let us continue the discussion there.